opa-manager-operator/tests/integration/test_charm.py

```diff
@@ -0,0 +1,113 @@
+import logging
+from pathlib import Path
+import os
+import time
+import pytest
+import yaml
+import re
+from kubernetes import client, config
+from kubernetes.client.rest import ApiException
+from jinja2 import Template
+
+log = logging.getLogger(__name__)
+
+meta = yaml.safe_load(Path("metadata.yaml").read_text())
+
+KUBECONFIG = os.environ.get("TESTING_KUBECONFIG", "~/.kube/config")
+
+config.load_kube_config(KUBECONFIG)  # TODO: check how to use this with designated file
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test):
+    charm = await ops_test.build_charm(".")
+    resources = {"gatekeeper-image": "openpolicyagent/gatekeeper:v3.2.3"}
+    for series in meta["series"]:
+        await ops_test.model.deploy(
+            charm, application_name="gatekeeper", series=series, resources=resources
+        )
+    await ops_test.model.wait_for_idle(wait_for_active=True, timeout=60 * 60)
```

Comment:
This test is failing here for me with a charm hook error:

```
application-gatekeeper: 13:34:40 ERROR unit.gatekeeper/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 172, in <module>
    main(OPAManagerCharm)
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/ops/main.py", line 406, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/ops/main.py", line 140, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/ops/framework.py", line 278, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/ops/framework.py", line 722, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/ops/framework.py", line 767, in _reemit
    custom_handler(event)
  File "./src/charm.py", line 152, in _on_start
    utils.create_k8s_object(os.environ["JUJU_MODEL_NAME"], k8s_object)
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/src/utils.py", line 111, in create_k8s_object
    ACTION_MAP.get(k8s_object.get("kind"), try_crd)(namespace, k8s_object, "create")
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/src/utils.py", line 24, in crud_pod_security_policy_with_api
    api_instance.create_pod_security_policy(body, pretty=True)
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/kubernetes/client/api/policy_v1beta1_api.py", line 203, in create_pod_security_policy
    return self.create_pod_security_policy_with_http_info(body, **kwargs)  # noqa: E501
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/kubernetes/client/api/policy_v1beta1_api.py", line 290, in create_pod_security_policy_with_http_info
    return self.api_client.call_api(
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/kubernetes/client/api_client.py", line 348, in call_api
    return self.__call_api(resource_path, method,
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/kubernetes/client/api_client.py", line 180, in __call_api
    response_data = self.request(
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/kubernetes/client/api_client.py", line 391, in request
    return self.rest_client.POST(url,
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/kubernetes/client/rest.py", line 274, in POST
    return self.request("POST", url,
  File "/var/lib/juju/agents/unit-gatekeeper-0/charm/venv/kubernetes/client/rest.py", line 233, in request
    raise ApiException(http_resp=r)
kubernetes.client.exceptions.ApiException: (403)
Reason: Forbidden
HTTP response headers: HTTPHeaderDict({'Audit-Id': '8c3399a2-608d-4109-b402-6c3a63b22294', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Content-Type-Options': 'nosniff', 'X-Kubernetes-Pf-Flowschem
a-Uid': 'd955bc68-d356-4a80-bf5f-3b488e906c21', 'X-Kubernetes-Pf-Prioritylevel-Uid': '759d6352-f87a-4532-9315-3bd33d5d31db', 'Date': 'Thu, 30 Sep 2021 18:34:40 GMT', 'Content-Length': '414'})
HTTP response body: {
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "podsecuritypolicies.policy is forbidden: User \"system:serviceaccount:k8s:gatekeeper-operator\" cannot create resource \"podsecuritypolicies\" in API group \"policy\" at the cluster scope",
  "reason": "Forbidden",
  "details": {
    "group": "policy",
    "kind": "podsecuritypolicies"
  },
  "code": 403
}

application-gatekeeper: 13:34:40 ERROR juju.worker.caasoperator.uniter.gatekeeper/0.operation hook "start" (via hook dispatching script: dispatch) failed: exit status 1

=================================== FAILURES ===================================
____________________________ test_build_and_deploy _____________________________
Traceback (most recent call last):
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-manager-operator/tests/integration/test_charm.py", line 29, in test_build_and_deploy
    await ops_test.model.wait_for_idle(wait_for_active=True, timeout=60 * 60)
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-manager-operator/.tox/integration/lib/python3.8/site-packages/juju/model.py", line 2329, in wait_for_idle
    _raise_for_status(errors, "error")
  File "/home/gkk/work/tasks/review/opa-gatekeeper-operators/opa-manager-operator/.tox/integration/lib/python3.8/site-packages/juju/model.py", line 2281, in _raise_for_status
    raise error_type("{}{} in {}: {}".format(
juju.errors.JujuUnitError: Unit in error: gatekeeper/0
```

I think because the gatekeeper-manager charm is trying to create a PodSecurityPolicy, but the test hasn't applied the gatekeeper-rb.yaml file yet. You might need to have the test apply the RBAC rules before deploying the charm.

Reply:
ack, thanks